### PR TITLE
Expose configurations as variants from local component metadata

### PR DESCRIPTION
**Pull request by jvandort**
The dependency resolution engine is largely variant-based. Modules contain components, and components contain variants. Variants are the node in the dependency graph.

The engine was not always like this. In prior iterations, nodes in the graph were configurations. Configurations have additional properties compared to variants, in that they have a hierarchy and can extend from one another. Furthermore, configurations are selected by name while variants are selected via attribute matching.

This commit largely finishes the transition of local components from the configuration type hierarchy to the variant type hierarchy.

- Renamed as many references of configuration to variant
- Local variant metadata and state types no longer extend from the configuration types
- Update error messages to reflect the fact that selection using `targetConfiguration` selects variants by their configuration name
- Allow variants to declare a configuration name that they are identified by. This is optional and paves the way to local variants that are not derived from configurations
- Create new IvyComponentGraphResolveState, which exposes configurations. Local components and other external components no longer expose configurations to the resolution engine. They can only expose variants
- Deprecate an edge-case where an ivy component can select a non-consumable configuration from a local component if the local component was substituted in place of a dependency of the ivy component.
- For now, we still expose a getConfigurationLegacy method on LocalComponentGraphResolveState, which is marked deprecated. This is to support the above-noted deprecated ivy behavior, and to support fetching the root node of a dependency graph. This second use case will be replaced in the near future.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/ivy/IvyDependencyMetadata.java [hunk 1]
```diff
@@ -69,7 +75,46 @@ protected GraphVariantSelectionResult selectLegacyConfigurations(
...
+            return new GraphVariantSelectionResult(Collections.singletonList(selected), false);
+        }
+
+        // Perhaps the legacy variant is present, but comes from a non-consumable configuration.
+        if (targetComponentState instanceof LocalComponentGraphResolveState) {
+            LocalComponentGraphResolveState localComponent = (LocalComponentGraphResolveState) targetComponentState;
+
+            // getConfigurationLegacy is a legacy mechanism and does _not_ check if the target variant comes from a consumable configuration
+            @SuppressWarnings("deprecation")
+            LocalVariantGraphResolveState legacyVariant = localComponent.getConfigurationLegacy(Dependency.DEFAULT_CONFIGURATION);
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/ivy/IvyDependencyMetadata.java:103]: This is one of two places we still call `getConfigurationLegacy`.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilder.java [hunk 2]
```diff
@@ -77,11 +77,16 @@ public RootComponentState toRootComponent(String configurationName) {
...
         LocalComponentGraphResolveState state = getComponentState(module, componentIdentifier);
-        ConfigurationGraphResolveState configuration = state.getConfiguration(configurationName);
-        if (configuration == null) {
+
+        // TODO: We should not ask the component for a resolvable configuration. Components should only
+        // expose variants -- which are by definition consumable only. Instead, we should create our own
+        // root variant and add it to a new one-off root component that holds only that root variant.
+        // The root variant should not live in a standard local component alongside other (consumable) variants.
+        @SuppressWarnings("deprecation")
+        LocalVariantGraphResolveState rootVariant = state.getConfigurationLegacy(configurationName);
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilder.java:86]: This is the other places `getConfigurationLegacy` is called.

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/describer/RequestedConfigurationNotFoundFailureDescriber.java [hunk 3]
```diff
@@ -35,6 +36,16 @@ public ConfigurationSelectionException describeFailure(RequestedConfigurationNot
     }
 
     private String buildConfigurationNotFoundFailureMsg(RequestedConfigurationNotFoundFailure failure) {
-        return String.format("A dependency was declared on configuration '%s' which is not declared in the descriptor for %s.", failure.getRequestedName(), failure.getRequestedComponentDisplayName());
+        String additionalMessage = "";
+        boolean isLocalComponent = failure.getRequestedComponentId() instanceof ProjectComponentIdentifier;
```
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/describer/RequestedConfigurationNotFoundFailureDescriber.java:40]: 🤔 I know we discussed changing this around a little to remove the `instanceof` check.  Just leave it for now, I'm working on the failure hierarchy and will adjust it then.

**Review by tresat — APPROVED** [review]

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 4]
```diff
@@ -337,6 +337,13 @@ plugins {
 
 We encourage you to use the https://plugins.gradle.org/plugin/com.gradle.develocity[latest released Develocity plugin version], even when you are using an older Gradle version.
 
+[[consuming_non_consumable_variants_from_ivy_component]]
+==== Deprecated consuming non-consumable configurations from Ivy
+
+In prior versions of Gradle, it was possible to consume non-consumable configurations of a project using published Ivy metadata.
+If an Ivy component's dependency was substituted for a project dependency, configurations in that project could be selected that were marked as non-consumable.
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:344]: +1

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantArtifactGraphResolveMetadata.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public interface LocalVariantArtifactGraphResolveMetadata extends VariantArtifactGraphResolveMetadata {
+
+    @Override
+    ImmutableList<? extends LocalComponentArtifactMetadata> getArtifacts();
+
+    /**
+     * Returns the "sub variants" of this variant.
+     *
+     * <p>This concept should disappear.</p>
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantArtifactGraphResolveMetadata.java:36]: this was copied. Reworeded.

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantArtifactGraphResolveMetadata.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Artifact metadata for local graph variants.
+ */
+public interface LocalVariantArtifactGraphResolveMetadata extends VariantArtifactGraphResolveMetadata {
+
+    @Override
+    ImmutableList<? extends LocalComponentArtifactMetadata> getArtifacts();
+
+    /**
+     * Returns the "sub variants" of this variant.
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantArtifactGraphResolveMetadata.java:34]: Reworded

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/internal/component/ResolutionFailureHandlerIntegrationTest.groovy [hunk 7]
```diff
@@ -225,9 +225,9 @@ class ResolutionFailureHandlerIntegrationTest extends AbstractIntegrationSpec {
         failure.assertHasDescription("Could not determine the dependencies of task ':forceResolution'.")
         failure.assertHasCause("Could not resolve all dependencies for configuration ':resolveMe'.")
         failure.assertHasCause("Could not resolve project :.")
-        assertFullMessageCorrect("""     Required by:
+        assertFullMessageCorrect("""Required by:
          project :
-      > A dependency was declared on configuration 'absent' which is not declared in the descriptor for project :.""")
+      > A dependency was declared on configuration 'absent' of 'project :' but no variant with that configuration name exists. The requested configuration is either not present in the target project or the named configuration is not consumable.""")
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/internal/component/ResolutionFailureHandlerIntegrationTest.groovy:230]: Updated wording

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ConfigurationBuildDependenciesIntegrationTest.groovy [hunk 8]
```diff
@@ -231,12 +231,14 @@ class ConfigurationBuildDependenciesIntegrationTest extends AbstractHttpDependen
         fails("useCompileConfiguration")
         failure.assertHasDescription("Could not determine the dependencies of task ':useCompileConfiguration'.")
         failure.assertHasCause("Could not resolve all dependencies for configuration ':compile'.")
-        failure.assertHasCause("Selected configuration 'default' on 'project :child' but it can't be used as a project dependency because it isn't intended for consumption by other components.")
+        failure.assertHasCause("""Unable to find a matching variant of project :child:
+  - No variants exist.""")
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ConfigurationBuildDependenciesIntegrationTest.groovy:235]: This is a very rare case. I think this is fine for now.

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**